### PR TITLE
pkg/envoy/ads: add error codes

### PR DESCRIPTION
**Pull request by jaellio**
**Description**:

Adds error codes for errors related to Envoy ADS.

Part of #2866

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [X] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? No
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change? No

### CODE REVIEW

File: pkg/errcode/errcode.go [hunk 1]
```diff
@@ -160,6 +160,69 @@ const (
 	ErrMeshConfigMarshaling
 )
 
+// Range 5000-5500 reserved for errors related to Envoy
```
Comment (allenlsy) [inline pkg/errcode/errcode.go:163]: Do you want to make it more specific, like "related to Envoy xDS server/control plane"?

File: pkg/envoy/ads/response.go [hunk 2]
```diff
@@ -73,7 +74,8 @@ func (s *Server) sendResponse(proxy *envoy.Proxy, server *xds_discovery.Aggregat
 		// Generate the resources for this request
 		resources, err := s.getTypeResources(proxy, finalReq)
 		if err != nil {
-			log.Error().Err(err).Msgf("Creating %s update for Proxy %s", typeURI.Short(), proxy.GetCertificateCommonName())
+			log.Error().Err(err).Str(errcode.Kind, errcode.ErrGeneratingReqResource.String()).
+				Msgf("Creating %s update for Proxy %s", typeURI.Short(), proxy.GetCertificateCommonName())
```
Comment (shashankram) [inline pkg/envoy/ads/response.go:78]:
```suggestion
				Msgf("Error generating response for typeURI: %s, proxy %s, typeURI.Short(), proxy.String())
```

File: pkg/envoy/ads/stream.go [hunk 3]
```diff
@@ -148,8 +151,9 @@ func shouldPushUpdate(proxy *envoy.Proxy) bool {
 	// In ADS, CDS and LDS will come first in all cases. Only allow an control-plane-push update push if
 	// we have sent either to the proxy already.
 	if proxy.GetLastSentNonce(envoy.TypeLDS) == "" && proxy.GetLastSentNonce(envoy.TypeCDS) == "" {
-		log.Error().Msgf("Proxy %s: LDS and CDS unrequested yet, waiting for first request for this proxy to be responded to",
-			proxy.String())
+		log.Error().Str(errcode.Kind, errcode.ErrProxyNotReady.String()).
```
Comment (shashankram) [inline pkg/envoy/ads/stream.go:154]: `ErrProxyNotReady` seems ambiguous here. Could you change this to `ErrUnexpectedXDSRequest` and update the description for the same.

File: pkg/envoy/ads/stream.go [hunk 4]
```diff
@@ -181,8 +185,9 @@ func respondToRequest(proxy *envoy.Proxy, discoveryRequest *xds_discovery.Discov
 	// Parse TypeURL of the request
 	typeURL, ok := envoy.ValidURI[discoveryRequest.TypeUrl]
 	if !ok {
-		log.Error().Msgf("Proxy %s: Unknown/Unsupported URI: %s",
-			proxy.String(), discoveryRequest.TypeUrl)
+		log.Error().Str(errcode.Kind, errcode.ErrInvalidURI.String()).
```
Comment (shashankram) [inline pkg/envoy/ads/stream.go:188]: s/ErrInvalidURI/ErrXDSInvalidTypeURI

File: pkg/errcode/errcode.go [hunk 5]
```diff
@@ -160,6 +160,69 @@ const (
...
+const (
+	// ErrMarshallingAccessLog indicates the Envoy AccessLog struct created could not be marshalled
+	ErrMarshallingAccessLog ErrCode = iota + 5000
```
Comment (shashankram) [inline pkg/errcode/errcode.go:166]: No need for a separate error code for each type when it comes to marshaling. A generic error code `ErrMarshallingXDSResource` should suffice.

File: pkg/errcode/errcode.go [hunk 6]
```diff
@@ -160,6 +160,69 @@ const (
...
+
+	// ErrGRPCStreamClosedByProxy indicates the gRPC stream was closed by the proxy
+	ErrGRPCStreamClosedByProxy
+
+	// ErrorProxyNotReady indicates that a proxy has not completed its init phase and is not ready to
+	// receive updates
+	ErrProxyNotReady
+
+	// ErrInvalidURI indicates the TypeURL of the discovery request is invalid
+	ErrInvalidURI
```
Comment (shashankram) [inline pkg/errcode/errcode.go:220]: ErrInvalidTypeURI

File: pkg/envoy/ads/stream.go (hunk 3 above)
Comment (jaellio) [reply pkg/envoy/ads/stream.go:154]: Would `ErrProxyWaitingForInitialUpdate` work?

File: pkg/errcode/errcode.go (hunk 6 above)
Comment (jaellio) [reply pkg/errcode/errcode.go]: Updated to `ErrInvalidXDSTypeURI`

File: pkg/envoy/ads/stream.go (hunk 4 above)
Comment (jaellio) [reply pkg/envoy/ads/stream.go]: Changed errcode to `ErrInvalidXDSTypeURI`

File: pkg/envoy/ads/stream.go (hunk 3 above)
Comment (jaellio) [reply pkg/envoy/ads/stream.go]: Update to `ErrUnexpectedXDSRequest`

**Review by shashankram — APPROVED** [review]

**Pull request merged**